#### Topic:      Prework Solutions
#### Cohort:   03_Spring2015
#### Date:        01-Apr-2015
#### Author:     Reshama Shaikh

Prework: Required Exercises

https://github.com/datascopeanalytics/metis-data-science-bootcamp-prework/blob/master/exercises.md

Source:     Think Stats (section Using the Code), there is some accompanying code and data. You can get these from the Think Stats repository.
https://github.com/AllenDowney/ThinkStats2

------------------------------------------------------------------------------- 

####Note:

Step1.  Create a directory on your computer.  Below is an example:

/Users/reshamashaikh/ds/metis/metisgh/

Step2.  use GitHub to pull this repo to your computer

git clone https://github.com/AllenDowney/ThinkStats2.git

Step3.  Put your ipython notebook in this directory (that way, it can pull the needed dependencies):  

/Users/reshamashaikh/ds/metis/metisgh/ThinkStats2/code/

(content will match:  https://github.com/AllenDowney/ThinkStats2/tree/master/code )

Step4.  Call your prework solutions notebook:  

0_S_Prework_myname.ipynb

##Table of Contents 
[Exercise01](#1) 

(Think Stats Ex 2.4) 

Using the variable totalwgt_lb, investigate whether first babies are lighter or heavier than others. Compute Cohen’s d to quantify the difference between the groups. How does it compare to the difference in pregnancy length?
 

[Exercise02](#2)

(Think Stats Ex 3.1) 

Something like the class size paradox appears if you survey children and ask how many children are in their family. Families with many children are more likely to appear in your sample, and families with no children have no chance to be in the sample.

Use the NSFG respondent variable NUMKDHH to construct the actual distribution for the number of children under 18 in the household.
Now compute the biased distribution we would see if we surveyed the children and asked them how many children under 18 (including themselves) are in their household.

Plot the actual and biased distributions, and compute their means. As a starting place, you can use chap03ex.ipynb. This is an ipython notebook from the ThinkStats2 repository.


[Exercise03](#3)

(Think Stats Ex 4.2) 

The numbers generated by random.random are supposed to be uniform between 0 and 1; that is, every value in the range should have the same probability.
Generate 1000 numbers from random.random and plot their PMF and CDF. Is the distribution uniform?



[Exercise04](#4)

(Think Stats Ex 7.1) 

Using data from the NSFG, make a scatter plot of birth weight versus mother’s age. Plot percentiles of birth weight versus mother’s age. Compute Pearson’s and Spearman’s correlations. How would you characterize the relationship between these variables?

[Exercise05](#5)

(Think Stats Ex 8.2)

Suppose that you draw a sample with size n = 10 from an exponential distribution with λ = 2. Simulate this experiment 1000 times and plot the sampling distribution of the estimate L. Compute the standard error of the estimate and the 90% confidence interval.

Repeat the experiment with a few different values of n and make a plot of standard error versus n.


[Exercise06](#6)

(Think Stats Ex 2.1 Bayes)

The cookie problem is a problem discussed in sections 1.3, 2.2 and 2.3 of Think Bayes. Solve the following problem. In Section 2.3 I said that the solution to the cookie problem generalizes to the case where we draw multiple cookies with replacement. But in the more likely scenario where we eat the cookies we draw, the likelihood of each draw depends on the previous draws.

Modify the solution in this chapter to handle selection without replacement. Hint: add instance variables to Cookie to represent the hypothetical state of the bowls, and modify Likelihood accordingly. You might want to define a Bowl object.


-----------------------------------------------------------------------------------------

### <a id='1'></a> Exercise 1 
(Think Stats Ex 2.4)

a) Using the variable totalwgt_lb, investigate whether first babies are lighter or heavier than others. 

b) Compute Cohen’s d to quantify the difference between the groups. 

c) How does it compare to the difference in pregnancy length?

Note:  Cohen's d is an effect size used to indicate the standardised difference between two means. It can be used, for example, to accompany reporting of t-test and ANOVA results. 

In [81]:
# set working directory
path_data = "/Users/reshamashaikh/_ds/metis/metisgh/ThinkStats2"

In [82]:
import nsfg
df = nsfg.ReadFemPreg()
print "Length of df: ", len(df)
df.head()

Length of df:  13593


caseid  pregordr  howpreg_n  howpreg_p  moscurrp  nowprgdk  pregend1  \
0       1         1        NaN        NaN       NaN       NaN         6   
1       1         2        NaN        NaN       NaN       NaN         6   
2       2         1        NaN        NaN       NaN       NaN         5   
3       2         2        NaN        NaN       NaN       NaN         6   
4       2         3        NaN        NaN       NaN       NaN         6   

   pregend2  nbrnaliv  multbrth    ...     laborfor_i  religion_i  metro_i  \
0       NaN         1       NaN    ...              0           0        0   
1       NaN         1       NaN    ...              0           0        0   
2       NaN         3         5    ...              0           0        0   
3       NaN         1       NaN    ...              0           0        0   
4       NaN         1       NaN    ...              0           0        0   

       basewgt  adj_mod_basewgt      finalwgt  secu_p  sest  cmintvw  \
0  3410.389399      3869.349602   6448.271112       2     9      NaN   
1  3410.389399      3869.349602   6448.271112       2     9      NaN   
2  7226.301740      8567.549110  12999.542264       2    12      NaN   
3  7226.301740      8567.549110  12999.542264       2    12      NaN   
4  7226.301740      8567.549110  12999.542264       2    12      NaN   

   totalwgt_lb  
0       8.8125  
1       7.8750  
2       9.1250  
3       7.0000  
4       6.1875  

[5 rows x 244 columns]

In [83]:
firsts = df[df.birthord==1]
others = df[df.birthord>1]
print "len of firsts: ", len(firsts)
print "len of others: ", len(others)

len of firsts:  4413
len of others:  4735


In [84]:
print "firsts total birth wt (mean): ", firsts.totalwgt_lb.mean()
print "others total birth wt (mean): ", others.totalwgt_lb.mean()

firsts total birth wt (mean):  7.20109443044
others total birth wt (mean):  7.32585561497


####Question: 1a) Using the variable totalwgt_lb, investigate whether first babies are lighter or heavier than others.
####Answer:   
####First babies are lighter than others (7.20 vs 7.33 lbs)

In [85]:
#Import some packages:
import pandas as pd
import matplotlib as mpl
from matplotlib import pyplot as plt
import numpy as np
import math

#***1.) Thinkstats exercise 2.4:
#Using the variable totalwgt_lb, investigate whether first babies are lighter
#or heavier than others. Compute Cohen's d to quantify the difference between 
#groups. How does it compare to the difference in pregnancy length?

#Change working directory to location of ThinkStats files and import data:
import os
os.chdir("/Users/ilya/ThinkStats2/code")
import nsfg
df = nsfg.ReadFemPreg()

#Remove non-live births data. Outcome == 1 means live birth.
live = df[df.outcome == 1]

#Pull apart first babies data from non-first babies data:
firsts = live[live.birthord == 1]
others = live[live.birthord != 1]

#Compute and print summary stats:
print firsts.totalwgt_lb.mean()
print firsts.totalwgt_lb.std()
print others.totalwgt_lb.mean()
print others.totalwgt_lb.std()

#Check for obvious data entry mistakes (over 25lbs is reasonable cutoff):
print firsts[firsts.totalwgt_lb > 25]
print others[others.totalwgt_lb > 25]

#Create histogram of all live birthweights:
live.totalwgt_lb.hist()
plt.xlabel('birthweights')
plt.ylabel('frequency')
plt.title('Distribution of live birthweights, n = %d' % len(live.totalwgt_lb))
plt.show()

#Compute Cohen's d to quantify the difference between groups:
def cohens_d(series1, series2):
    """Computes cohen's d effect size"""

    mean1, mean2 = series1.mean(), series2.mean()
    diff = mean1 - mean2
    var1, var2 = series1.var(), series2.var()
    n1, n2 = len(series1), len(series2)

    pooled_var = (n1 * var1 + n2 * var2) / (n1 + n2)
    d = diff / math.sqrt(pooled_var)

    return d

print cohens_d(firsts.totalwgt_lb, others.totalwgt_lb)

#Cohen's d is -.09. First-born babies weigh marginally less than ones born 
#later. The effect here is similar to the difference in pregnancy lengths, and 
#like the difference in pregnancy lengths, is tiny.


####Answer:  

####Cohen's D is:  -0.0887

c) How does Cohen's D compare to the difference in pregnancy length? 

In [86]:
print "Cohen's D for birth weight is:      x.xxxx"
print "Cohen's D for pregnancy length is:  x.xxxx"

Cohen's D for birth weight is:      x.xxxx
Cohen's D for pregnancy length is:  x.xxxx


####Interpretation
This site has a good interpretation of Cohen's D, effect size:
http://www.uccs.edu/lbecker/effect-size.html

[fill in ] 

-------------------------------------------------------------------------------------------

### <a id='2'></a> Exercise 2 
(Think Stats Ex 3.1)

Something like the class size paradox appears if you survey children and ask how many children are in their family. Families with many children are more likely to appear in your sample, and families with no children have no chance to be in the sample.

Use the NSFG respondent variable NUMKDHH to construct the actual distribution for the number of children under 18 in the household.

Now compute the biased distribution we would see if we surveyed the children and asked them how many children under 18 (including themselves) are in their household.

Plot the actual and biased distributions, and compute their means. As a starting place, you can use chap03ex.ipynb. This is an ipython notebook from the ThinkStats2 repository.

In [ ]:
#***2.) Thinkstats exercise 3.1:
#Use the NSFG variable NUMKDHH to construct the actual distribution for the 
#number of children under 18 in households. Then compute the biased 
#distribution we'd see if we surveyed children and asked them how many children
#including themselves are in the household. Plot actual and biased 
#distributions and compute their means. 

#import data and thinkstats package for pmfs:
import thinkstats2 as ts
import thinkplot as tp
import chap01soln
dat = chap01soln.ReadFemResp()

#Make and display probability mass function:
pmf = ts.Pmf(dat.numkdhh)
print pmf
tp.hist(pmf)
plt.xlim([-.5,5.5])
plt.xlabel('Number of children')
plt.ylabel('Probability mass')
plt.title('Probability mass distribution of children per household')
plt.show()

#Define function that calculates biased pmf and normalizes the probabilities:
def BiasPmf(pmf, label = ''):
    """Returns the Pmf with oversampling proportional to value."""

    new_pmf = pmf.Copy(label = label)
    
    for x, p in pmf.Items():
        new_pmf.Mult(x, x)

    new_pmf.Normalize()

    return new_pmf

#Make biased pmf using above function:
biased_pmf = BiasPmf(pmf, 'Biased_pmf')
print biased_pmf

#Plot biased and unbiased pmfs on same axes:
tp.Hist(pmf, align = 'right', width = .35, label = 'Unbiased')
tp.Hist(biased_pmf, align = 'left', width = .35, label = 'Biased')
plt.xlim([-.5,5.5])
plt.xlabel('Number of children')
plt.ylabel('Probability mass')
plt.title('Biased and unbiased probability mass distributions of \nchildren \
per household')
tp.Show()

#Compute means of the two pmfs:
print 'The unbiased mean children per household is %f' % pmf.Mean()
print 'The biased mean children per household is %f'  % biased_pmf.Mean()



-------------------------------------------------------------------------------------------

### <a id='3'></a> Exercise 3 
(Think Stats Ex 4.2)

The numbers generated by random.random are supposed to be uniform between 0 and 1; that is, every value in the range should have the same probability.
Generate 1000 numbers from random.random and plot their PMF and CDF. Is the distribution uniform?

In [87]:
import thinkstats2
#Import random module and generate numbers into a list:
import random
numbers = [random.random() for _ in range(1000)]

#Plot PMF:
pmf = ts.Pmf(numbers)
tp.Pmf(pmf, width = 0.01)
plt.xlabel('Number')
plt.ylabel('Probability mass')
plt.title('Probability mass distribution of 1000 numbers generated with random.random')
tp.Show()

#Plot CDF:
cdf = ts.Cdf(numbers)
plt.title('CDF for 1000 supposedly random numbers')
plt.xlabel('Number')
plt.ylabel('Probability mass')
tp.Cdf(cdf)
tp.Show()

#Yes, the distribution appears uniform.

####Answer:
Yes, distribution looks uniform

-------------------------------------------------------------------------------------------

### <a id='4'></a> Exercise 4 
(Think Stats Ex 7.1)

Using data from the NSFG, make a scatter plot of birth weight versus mother’s age. Plot percentiles of birth weight versus mother’s age. Compute Pearson’s and Spearman’s correlations. How would you characterize the relationship between these variables?

In [88]:
import nsfg

#Import and clean data
dat = nsfg.ReadFemPreg()
dat = dat.dropna(subset = ['totalwgt_lb'])


#dat = dat.dropna(subset = ['totalwgt_lb', ''])

#Scatterplot birthweight against mother's age:
age, weight = dat.agepreg, dat.totalwgt_lb
tp.Scatter(age, weight, alpha = .2)
plt.title('Mother\'s age vs baby birthweight')
tp.Show(xlabel = 'Mother\'s age', ylabel = 'Birth weight',
    axis = [9,46,-1,16])

#Hexplot birthweight against mother's age because there are so many datapoints
#in the middle.
tp.HexBin(age, weight)
plt.title('Mother\'s age vs baby birthweight')
tp.Show(xlabel = 'Mother\'s age', ylabel = 'Birth weight')

#Bin data so that you can calculate CDF for each bin seperately:
bins = np.arange(0, 55, 5)
indices = np.digitize(dat.agepreg, bins)
groups = dat.groupby(indices)

#Compute mean age and CDF of birthweight:
ages = [group.agepreg.mean() for i, group in groups][1:-1]
cdfs = [ts.Cdf(group.totalwgt_lb) for i, group in groups]

#Plot percentiles of weight vs height:
for percent in [75, 50, 25]:
    weights = [cdf.Percentile(percent) for cdf in cdfs[1:-1]]
    label = '%dth' % percent
    tp.Plot(ages, weights, label = label)

plt.title('Mother\'s age vs baby birthweight by percentile')
tp.Show()

#Compute Pearson's correlation:
from scipy import stats
print stats.pearsonr(age, weight)

#Compute Spearman's correlation:
print stats.spearmanr(age, weight)

#Pearson's r = .07, Spearman's r = .09. Both of these are small effects. The 
#relationship is weakly positive-- as mother's age increases, so does baby
#birthweight.


####Answer:
For the Pearson r correlation, both variables should be normally distributed.  Other assumptions include linearity and homoscedasticity.  Linearity assumes a straight line relationship between each of the variables in the analysis and homoscedasticity assumes that data is normally distributed about the regression line.

Spearman rank correlation test does not make any assumptions about the distribution. 

####Conclusions:
Data seem to be  ....

-----------------------------------------------------------------------------------------

### <a id='5'></a> Exercise 5
(Think Stats Ex 8.2)

Suppose that you draw a sample with size n = 10 from an exponential distribution with λ = 2. Simulate this experiment 1000 times and plot the sampling distribution of the estimate L. Compute the standard error of the estimate and the 90% confidence interval.

Repeat the experiment with a few different values of n and make a plot of standard error versus n.

In [89]:
import thinkstats2

#Define root mean square error calculator:
def RMSE(estimates, actual):
    e2 = [(estimate - actual)**2 for estimate in estimates]
    mse = np.mean(e2)
    return math.sqrt(mse)

#Define function to run simulation a bunch of times and return rmse and list of
#estimates.
def estimate_exp(n, m):
    lam = 2
    estimates = []
    for _ in range(m):
        xs = np.random.exponential(1.0/lam, n)
        L = 1 / np.mean(xs)
        estimates.append(L)
    return estimates, RMSE(estimates, lam)

#Define function that calculates percentile from percentile rank:
def PercentileRank(values, your_value):
    count = 0
    for value in values:
        if value <= your_value:
            count += 1

    return 100 * count / len(values)

def Percentile(values, percentile_rank):
    values.sort()
    for val in values:
        if PercentileRank(values, val) >= percentile_rank:
            return val

#Run simulation 1000 times, with 7 samples from the exp distribution per time
#and return the estimates and root mean square error:
estimates_list, rmse = estimate_exp(10, 1000)

#Compute 95% confidence interval:
low, high = Percentile(estimates_list, 5), Percentile(estimates_list, 95)

#With sample size 10, the confidence interval is 1.28, 3.73 and the RMSE = .84.

#Plot the sampling distribution of the lambda estimator L with confidence int:
cdf = ts.Cdf(estimates_list)
plt.title('CDF for 1000 estimates of lambda')
plt.xlabel('Estimate')
plt.ylabel('Probability mass')
plt.axvline(low, color = 'red', linestyle = 'solid', alpha = .3)
plt.axvline(high, color = 'red', linestyle = 'solid', alpha = .3)
tp.Cdf(cdf)
tp.Show()

#Repeat experiment with values of n between 10 and 1000 that end in 0:
rmse_list = []
for n in range(10, 1000, 10):
    rmse_list.append(estimate_exp(n, 1000)[1])

#Plot standard error (RMSE) vs n:
tp.Scatter(range(10,1000,10), rmse_list)
plt.title('RMSE for different sample sizes')
tp.Show(xlabel = 'number of items in sample', ylabel = 'RMSE', 
    axis = [0,1000,0,1])

-----------------------------------------------------------------------------------------

### <a id='6'></a> Exercise 6
(Think Stats Ex 2.1 Bayes)

The cookie problem is a problem discussed in sections 1.3, 2.2 and 2.3 of Think Bayes. Solve the following problem. In Section 2.3 I said that the solution to the cookie problem generalizes to the case where we draw multiple cookies with replacement. But in the more likely scenario where we eat the cookies we draw, the likelihood of each draw depends on the previous draws.

Modify the solution in this chapter to handle selection without replacement. Hint: add instance variables to Cookie to represent the hypothetical state of the bowls, and modify Likelihood accordingly. You might want to define a Bowl object.

In [90]:
# get this file from:  https://github.com/AllenDowney/ThinkBayes/blob/master/thinkbayes.py
from thinkbayes import Pmf

In [ ]:
#I can't figure out how to change the thinkbayes book's code, so I'm writing my
#own. To test it, make an object of the bayesian_cookies class and run the last
#function, called cookie_order() on the object, passing a list of cookies to eat
#in the order in which they're eaten. Vanilla cookies should be called "van" and
#chocolate ones should be called "choc". At the end, I put an example of 
#cookie_order() processing 3 cookies.

class bayesian_cookies():

    def __init__(self):
        pass

    bowls = {'b1': {'choc': 10, 'van': 30}, 'b2': {'choc': 20, 'van': 20}}

    def p_cookie_given_bowl(self, cookie, bowl, not_cookie):
        """This returns the likelihood, which is the probability of the data
        under the hypothesis"""

        return self.bowls[bowl][cookie] / float(self.bowls[bowl][cookie] + 
            self.bowls[bowl][not_cookie])

    p_bowl = [.5, .5] #This is the prior probability of choosing either bowl.

    def p_cookie(self, cookie, not_cookie):
        """Returns the normalizing constant, which is the probability
        of the data under ANY hypothesis."""

        p_v_b1 = self.bowls['b1'][cookie] / float(self.bowls['b1'][cookie] + 
            self.bowls['b1'][not_cookie])
        p_v_b2 = self.bowls['b2'][cookie] / float(self.bowls['b2'][cookie] + 
            self.bowls['b2'][not_cookie])
        return (p_v_b1 * self.p_bowl[0]) + (p_v_b2 * self.p_bowl[1])
        # cookie_sum = self.bowls['b1'][cookie] + self.bowls['b2'][cookie]
        # not_cookie_sum = self.bowls['b1'][not_cookie] + \
        #     self.bowls['b2'][not_cookie]
        # return cookie_sum / float(cookie_sum + not_cookie_sum)

    def bayes_theorem(self, p_b_given_a, p_a, p_b):
        """This defines bayes theorem, but it needs the numbers inputted"""

        return (p_b_given_a * p_a) / p_b

    #Probability of bowl 1 given a vanilla cookie:
    def p_b1_given_van(self):
        """This passes the relevant numbers through bayes_theorem to find out
        the prob of bowl 1 given a vanilla cookie"""

        return self.bayes_theorem(self.p_cookie_given_bowl('van', 'b1', 
            'choc'), self.p_bowl[0], self.p_cookie('van', not_cookie = 'choc'))

    #Prob of bowl 2 given a vanilla cookie:
    def p_b2_given_van(self):
        """This passes the relevant numbers through bayes_theorem to find out
        the prob of bowl 2 given a vanilla cookie"""
        return 1 - self.p_b1_given_van()
        # self.bayes_theorem(self.p_cookie_given_bowl('van', 'b2', 
        #     'choc'), self.p_bowl[1], self.p_cookie('van', not_cookie = 'choc'))

    #Prob of bowl 1 given a chocolate cookie:
    def p_b1_given_choc(self):
        return self.bayes_theorem(self.p_cookie_given_bowl('choc', 'b1', 
            'van'), self.p_bowl[0], self.p_cookie('choc', not_cookie = 'van'))

    #Prob of bowl 2 given a chocolate cookie:
    def p_b2_given_choc(self):
        return 1 - self.p_b1_given_choc

    def eat_cookie_and_update(self, cookie_type):
        if cookie_type == 'van':
            # If you chose a vanilla cookie, update the number of vanilla
            # cookies and recalculate the probability of choosing one of them,
            # and update the probability of choosing a chocolate one
            self.p_bowl = [self.p_b1_given_van(), self.p_b2_given_van()]
            print "The probabilities of a %s cookie coming from bowls 1 and 2,\
                respectively: %r" % (cookie_type, self.p_bowl)
        
        if cookie_type == 'choc':
            self.p_bowl = [self.p_b1_given_choc(), self.p_b2_given_choc()]
            print "The probabilities of a %s cookie coming from bowls 1 and 2,\
                respectively: %r" % (cookie_type, self.p_bowl)
        
        self.bowls['b1'][cookie_type] -= 1
        self.bowls['b2'][cookie_type] -= 1
        print "The bowl breakdown now is: %r" % self.bowls

    def cookie_order(self, order):
        """Takes a list of cookies in the order in which they're eaten and
        returns the probabilities of each cookie coming from each of the
        two bowls."""

        for cookie in order:
            self.eat_cookie_and_update(cookie)

eater = bayesian_cookies()
print eater.cookie_order(['van','choc','van'])